app/forms/hub/take_action_form.rb

```diff
@@ -40,14 +48,86 @@ def contact_method_help_text
       end
     end
 
+    def take_action!
+      tax_return = @client.tax_returns.find_by(id: tax_return_id)
+      raise TakeActionFormError.new("Can't update tax return unrelated to current client.") unless tax_return.present?
+      raise TakeActionFormError.new("Can't initiate status change to current status.") if status == tax_return.status
+
+      if tax_return.update!(status: status)
+        @action_list << I18n.t("hub.clients.update_take_action.flash_message.status")
+        SystemNote.create_status_change_note(@current_user, tax_return)
+        send_message if message_body.present?
+        create_note if internal_note_body.present?
+      end
+    end
+
     def self.permitted_params
-      [:tax_returns, :locale, :message_body, :contact_method, :internal_note_body, { tax_returns: {} }]
+      [:tax_return_id, :status, :locale, :message_body, :contact_method, :internal_note_body]
     end
 
     private
 
+    def send_message
+      case contact_method
+      when "email"
+        send_email(message_body, subject_locale: locale)
+        @action_list << I18n.t("hub.clients.update_take_action.flash_message.email")
+      when "text_message"
+        send_text_message(message_body)
+        @action_list << I18n.t("hub.clients.update_take_action.flash_message.text_message")
+      end
+    end
+
+    def create_note
+      Note.create!(
+          body: internal_note_body,
+          client: @client,
+          user: @current_user
+      )
+      @action_list << I18n.t("hub.clients.update_take_action.flash_message.internal_note")
+    end
+
+    def current_user
```

Comment:
This makes the `Model` into sort-of a controller. I wonder if this is evidence that MessageSending is the wrong abstraction, and perhaps `send_email` needs to live on `OutgoingEmail`, or perhaps somewhere else. Non-blocking, but curious for your take.

Reply:
agreed! I think we should have those be class methods on the model, or some kind of service object instead of a controller mixin -- I think we're going to be wanting to send emails /messages from a variety of places, not just controllers